sql/src/main/java/io/crate/analyze/Relations.java

```diff
@@ -35,4 +38,36 @@
     static Collection<? extends Path> namesFromOutputs(List<Symbol> outputs) {
         return Lists.transform(outputs, Symbols::pathFromSymbol);
     }
+
+    /**
+     * Promotes the relation to a QueriedRelation by applying the QuerySpec.
+     *
+     * <pre>
+     * TableRelation -> QueriedTable
+     * QueriedTable  -> QueriedSelect
+     * QueriedSelect -> nested QueriedSelect
+     * </pre>
+     *
+     * If the result is a QueriedTable it is also normalized.
+     */
+    static QueriedRelation applyQSToRelation(Functions functions,
+                                             TransactionContext transactionContext,
+                                             AnalyzedRelation relation,
+                                             QuerySpec querySpec) {
+        QueriedRelation newRelation;
+        if (relation instanceof DocTableRelation) {
```

Comment:
using a Visitor would require a Context class because there are 3 arguments that are required in addition to the relation (functions, transactionContext, querySpec). `instanceof` should be cheaper.